Sources/Datadog/Core/System/Time/DateCorrector.swift

```diff
@@ -38,7 +38,11 @@ internal class DateCorrector: DateCorrectorType {
         self.serverDateProvider = serverDateProvider
         serverDateProvider.synchronize(
             with: DateCorrector.datadogNTPServers.randomElement()!, // swiftlint:disable:this force_unwrapping
-            completion: { offset in
+            completion: { [weak self] offset in
+                guard let _ = self else {
+                    return
+                }
+
```

Comment:
This change is to address flaky execution of

https://github.com/DataDog/dd-sdk-ios/blob/181359bd655b894041702332c26fa05b2d451990/Tests/DatadogTests/Datadog/Core/Upload/DataUploadWorkerTests.swift#L230-L235

Problem was that `DateCorrector` is expected to perform real NTP sync in many high-level tests (e.g. most tests in [`DatadogTests`](https://github.com/DataDog/dd-sdk-ios/blob/181359bd655b894041702332c26fa05b2d451990/Tests/DatadogTests/Datadog/DatadogTests.swift#L10)). This closure might leak from test execution and send NTP sync log to global `userLogger`, causing bad value being recorded in other tests that depend on global `userLogger` mock, e.g.:

<img width="1138" alt="Screenshot 2022-01-10 at 10 19 07" src="https://user-images.githubusercontent.com/2358722/148773151-db7485ef-e685-4e3f-9b76-0c92b9398570.png">

The fix is to not leak this closure - and return it early when SDK was deallocated.

Reply:
Would this work? 
```swift
completion: { [weak userLogger] offset in
```
Not a change request, I'm not sure about lifecycle of the logger vs. DateCorrector 🤔

Reply:
`DateCorrector` is instantiated [only once for a given SDK instance](https://github.com/DataDog/dd-sdk-ios/blob/b4d285b0820fb83d561c492a8c740d6ba87307cf/Sources/Datadog/Datadog.swift#L177-L180) (from `Datadog.initialize()`).

I think more correct solution is to avoid doing **any work** in this `DateCorrector` closure if `self == nil`. From `self == nil` we know that `DateCorrector` was deallocated → so the parent `Datadog` object was deallocated → so no-one is interested in result of this work.

Capturing weak reference to `userLogger` *might* work too but it will be ignoring only a portion of the completion work related to `userLogger`. As no-one will use the result of `completion` work, I find this odd.

How does it sound?

Reply:
I guess the root cause is that `Kronos.Clock` is static as well as `userLogger`. I think it's best to fix it in [`ServerDateProvider`](https://github.com/DataDog/dd-sdk-ios/blob/master/Sources/Datadog/Core/System/Time/ServerDateProvider.swift#L35L46) and call the completion only if `self` exist, WDYT?

Reply:
> I think it's best to fix it in [`ServerDateProvider`](https://github.com/DataDog/dd-sdk-ios/blob/master/Sources/Datadog/Core/System/Time/ServerDateProvider.swift#L35L46) and call the completion only if `self` exist, WDYT?

Indeed! It depends on `self` already and clearly it runs its completion inconsitently when `self` is `nil`. Will do 👍

Reply:
Done ✅